src/app/components/wrapper-components/part-creation/part-creation.component.spec.ts

```diff
@@ -9,620 +12,715 @@ import { PartDAO } from 'src/app/dao/PartDAO';
+                void userDAO.update(UserMocks.CREATOR_AUTH_USER.userId, { state: 'offline' });
+                void partDAO.set('joinerId', PartMocks.INITIAL);
 
                 // When arriving on that component
-                testUtils.detectChanges();
-                tick(3000); // needs to be >2999
+                tick();
 
                 // Then the part and all its related data should be removed
-                expect(gameService.deletePart).toHaveBeenCalledWith('joinerId');
-                expect(joinerService.deleteJoiner).toHaveBeenCalledWith();
-                expect(chatService.deleteChat).toHaveBeenCalledWith('joinerId');
+                expect(gameService.deletePart).not.toHaveBeenCalled();
+                expect(joinerService.deleteJoiner).not.toHaveBeenCalled();
+                expect(chatService.deleteChat).not.toHaveBeenCalled();
+                component.stopSendingPresenceTokensAndObservingUsersIfNeeded();
             }));
         });
-        describe('Waiting for config', () => {
-            it('should reroute to server when game is cancelled', fakeAsync(async() => {
+        describe('Leaving', () => {
+            it('should remove yourself when leaving the room', fakeAsync(() => {
+                // Given a partCreation where user is but candidate
+                awaitComponentInitialisation();
+                expect(component.currentJoiner).toEqual(JoinerMocks.WITH_FIRST_CANDIDATE);
+
+                // When leaving the page (testing here by calling ngOnDestroy)
+                spyOn(joinerService, 'unsubscribe').and.callFake(() => {});
+                spyOn(joinerService, 'cancelJoining').and.callFake(async(user: MinimalUser) => {});
+                void component.ngOnDestroy();
+                tick();
+
+                // Then user's name taken out of the candidate list
+                expect(joinerService.cancelJoining).toHaveBeenCalledOnceWith(UserMocks.OPPONENT_MINIMAL_USER);
+            }));
+            it('should remove observedPart when leaving the room', fakeAsync(() => {
+                // Given a partCreation where user is but candidate
+                awaitComponentInitialisation();
+                expect(component.currentJoiner).toEqual(JoinerMocks.WITH_FIRST_CANDIDATE);
+
+                // When leaving the page (testing here by calling ngOnDestroy)
+                const authService: ConnectedUserService = TestBed.inject(ConnectedUserService);
+                spyOn(authService, 'removeObservedPart').and.callThrough();
+                spyOn(joinerService, 'unsubscribe').and.callFake(() => {});
+                spyOn(joinerService, 'cancelJoining').and.callFake(async(user: MinimalUser) => {});
+                void component.ngOnDestroy();
+                tick();
+
+                // Then observed part should be removed
+                expect(authService.removeObservedPart).toHaveBeenCalledOnceWith();
```

Comment:
mergeable avec le test précédent? Y'a que le Then qui change

Reply:
Oui mais je me suis dit que peut être c'est bien de les rendre un peu indépendant, mais fait!
J'ai en contre partie une question à poser...
Les spyOn et préparation de mock, ça devrait être du given ou du when ??

Reply:
hmm les spyOn ça dépend je dirais : idéalement on les mettrait avant le given, mais parfois c'est trop tôt donc faut le mettre avant le when... Mais c'est pas tellement quelque chose de phrasé ni dans le G ni dans le W

Reply:
J'dirais qu'ils sont plus dans le given que dans le when (given that a spy that and mock this, when doing, then spied one should blblb)